src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1174,21 +1182,42 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
         }
 
         /// <exception cref="UnsupportedSignatureContent">If the encoded parameter type is invalid.</exception>
-        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info)
+        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info, bool isReturn)
         {
+            var allowedRequiredModifiers = AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute;
+            if (isReturn)
+            {
+                // PROTOTYPE(init-only): can we make this more restrictive (ie. disallow aside from the return value of an instance setter)?
+                allowedRequiredModifiers |= AllowedRequiredModifierType.System_Runtime_CompilerServices_IsExternalInit;
+            }
+
             info.CustomModifiers = DecodeModifiersOrThrow(
                 ref signatureReader,
-                AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute,
+                allowedRequiredModifiers,
                 out SignatureTypeCode typeCode,
-                out bool inAttributeFound);
+                out AllowedRequiredModifierType inAttributeFound);
 
             if (typeCode == SignatureTypeCode.ByReference)
             {
                 info.IsByRef = true;
-                info.RefCustomModifiers = info.CustomModifiers;
-                info.CustomModifiers = DecodeModifiersOrThrow(ref signatureReader, AllowedRequiredModifierType.None, out typeCode, out _);
+
+                var refModifiers = info.CustomModifiers;
+                if (!refModifiers.IsDefault)
+                {
+                    refModifiers = refModifiers.WhereAsArray(m => IsAcceptedInAttributeModifierType(m.Modifier));
```

Comment:
>refModifiers = refModifiers.WhereAsArray(m => IsAcceptedInAttributeModifierType(m.Modifier)); [](start = 20, length = 93)

The change here doesn't make any sense. It is not appropriate to apply any filtering to custom modifiers or move them around as done below. We can either accept them without an error, or refuse to accept them. The specification says that modreq is applied to return type, therefore, if it is applied to a reference, it is an error situation and we should handle that the same way as we handling any modreq other than IsExternalInit on the return type below (I am referring to condition ```(inAttributeFound & ~AllowedRequiredModifierType.System_Runtime_CompilerServices_IsExternalInit) != 0```). Here, the only modreq that we can allow on ref is InAttribute, anything else is unsupported. When that is done we are decoding the next set of modifiers, those are actually applied to the type. We should answer the question whether we support ref-returning init properties. I guess not, but if we want to support them, then we need to allow IsExternalInit in that set.

Reply:
Thanks

---
In reply to: [416715729](https://github.com/dotnet/roslyn/pull/43275#discussion_r416715729) [](ancestors = 416715729)